File: IOS/Services/Registration/Validator.swift
//
//  Validator.swift
//  MP
//
//  Created by aleksej on 3/22/24.
//

import Foundation

typealias validatorClosure = (RegistrationForm) -> (isSuccessful: Bool, reason: String?)

class Validator
{
    static let namePredicate: NSPredicate = NSPredicate(format: "SELF MATCHES %@", "^(?!.*[\\s])[A-Za-z]{1,}$")
    static let emailPredicate: NSPredicate = NSPredicate(format: "SELF MATCHES %@", "^(?=.*[\\@\\.]).{5,}$")
    static let passwordPredicate: NSPredicate = NSPredicate(format: "SELF MATCHES %@", "^(?=.*[A-Z]).{8,}$")
    
    enum ValidatorType
    {
        case VT_Email, VT_Password, VT_FirstName, VT_LastName
    }
    
    static func getValidator(type: ValidatorType) -> validatorClosure
    {
        switch type {
        case .VT_Email:
            return validateEmail
        case .VT_Password:
            return validatePassword
        case .VT_FirstName, .VT_LastName:
            return validateName
        }
    }

    private static func validateEmail(form: RegistrationForm) -> (isSuccessful: Bool, reason: String?)
    {
        let result = emailPredicate.evaluate(with: form.email)
        return (result, result ? nil : "The email doesn't meet the requirements!")
    }

    private static func validatePassword(form: RegistrationForm) -> (isSuccessful: Bool, reason: String?)
    {
        let result = passwordPredicate.evaluate(with: form.password)
        return (result, result ? nil : "The password doesn't meet the requirements!")
    }
    
    private static func validateName(form: RegistrationForm) -> (isSuccessful: Bool, reason: String?)
    {
        let result = namePredicate.evaluate(with: form.firstName) && namePredicate.evaluate(with: form.lastName)
        return (result, result ? nil : "Names don't meet the requirements!")
    }
}

File: KOTLIN/src/main/java/by/bsuir/library/model/Validator.kt
package by.bsuir.library.model

typealias ValidatorClosure = (RegistrationForm) -> Pair<Boolean, String?>

enum class ValidatorType {
    EmailVT, PasswordVT, FNameVT, LNameVT
}

class Validator {
    companion object {
        fun getValidator(type: ValidatorType): ValidatorClosure {
            return when (type) {
                ValidatorType.EmailVT -> ::validateEmail
                ValidatorType.PasswordVT -> ::validatePassword
                ValidatorType.FNameVT -> ::validateName
                ValidatorType.LNameVT -> ::validateName
            }
        }

        private fun validateEmail(form: RegistrationForm): Pair<Boolean, String?> {
            val result = form.email != null && form.email!!.isNotEmpty() && !form.email!!.contains(Regex("\\s"))
            return Pair(result, if (result) null else "The email doesn't meet the requirements!")
        }

        private fun validatePassword(form: RegistrationForm): Pair<Boolean, String?> {
            val result = form.password != null && form.password!!.isNotEmpty() && !form.password!!.contains(Regex("\\s"))
            return Pair(result, if (result) null else "The password doesn't meet the requirements!")
        }

        private fun validateName(form: RegistrationForm): Pair<Boolean, String?> {
            val result = form.firstName != null && form.firstName!!.isNotEmpty() &&
                    form.lastName != null && form.lastName!!.isNotEmpty() && !form.firstName!!.contains(Regex("\\s")) && !form.lastName!!.contains(Regex("\\s"))
            return Pair(result, if (result) null else "Names don't meet the requirements!")
        }
    }
}